### ISSUE TITLE
Bug in cudnnSetTensor

### ISSUE THREAD

**Issue by blueberry** [jcuda/jcudnn#4]
Hi @jcuda 

It seems that there is a bug in the native JNI function that calls cudnnSetTensor.

Since I use this from Clojure, I can't show the pure Java code that reproduces this (since there is none) but here is the reasoning what I think the problem is and why I think that the problem is in JCuda rather than in my code.

The problem: When I call cudnnSetTensor, I always get  CUDNN_STATUS_BAD_PARAM. 

Per cuDNN documentation, it means that "One of the provided pointers is nil."

However, I am pretty sure that the parameters are OK. Not only that, but I tried all variations of device and host scalars, buffers, arrays and all types normally supported by JCuda. Always the same.

Now, why I think that my code is OK? There is a function with virtually identical signature: cudnnScaleTensor, which works OK, and returns SUCCESS.

Now, compare these two signatures:

```
cudnnStatus_t cudnnSetTensor(
    cudnnHandle_t                   handle,
    const cudnnTensorDescriptor_t   yDesc,
    void                           *y,
    const void                     *valuePtr)

cudnnStatus_t cudnnScaleTensor(
    cudnnHandle_t                   handle,
    const cudnnTensorDescriptor_t   yDesc,
    void                           *y,
    const void                     *alpha)
```

alpha
Input. Pointer in the host memory to a single value that all elements of the tensor will be scaled with.

valuePtr
Input. Pointer in host memory to a single value. All elements of the y tensor will be set to value[0]. 

This indicates that these functions should be called as (pseudocode):

```
cudnnSetTensor(handler, yDesc, jcuda-buffer-pointer, pointer-to-host-scalar)
```

HOWEVER, JCuda JNI code dealing with alpha and valuePtr is different for these two functions. 

https://github.com/jcuda/jcudnn/blob/7d63697dc48984ec337ae91b53ec18e2ac58d2ab/JCudnnJNI/src/JCudnn.cpp#L2697

https://github.com/jcuda/jcudnn/blob/7d63697dc48984ec337ae91b53ec18e2ac58d2ab/JCudnnJNI/src/JCudnn.cpp#L2637

The function that works well uses an extra step

```
PointerData *alpha_pointerData = initPointerData(env, alpha);
    if (alpha_pointerData == NULL)
    {
        return JCUDNN_STATUS_INTERNAL_ERROR;
    }
    alpha_native = (void *)alpha_pointerData->getPointer(env);
```

while the one that always raises an error from cuDNN just forwards the bare pointer to the cudnnSetTensor.

My hunch is that instead of the void pointer to HOST memory, it assumes that the pointer points to the device memory (which cuDNN assumes illegal and raises the exception).

I would gladly provide a fix, but I can't find enough documentation about JCuda JNI internals to clearly understand what is going o,n so I guess that I have to bug you about this :) Sorry :(

Am I on the right track with this?

**Comment by jcuda** [comment]
That's indeed a bug. It will be fixed with the next release (i.e. when I update JCuda for CUDA 10.2, and do the corresponding update/release for the then-latest CUDNN).

Thanks for pointing that out!

---

For the case that someone is curious, here's the reason for this bug (although it is somewhat embarrassing...) : 

In JCuda, pointers to host memory and pointers to device memory are treated differently. 

<sup>(One could go *very* far with describing the reasons for that - it's roughly related to the garbage collector, or more specifically, to the fact that JCuda is one of the few JNI libraries that always tried to support host memory that was *not* created as a direct `ByteBuffer`, but only as a plain, primitive Java array...)</sup>

This worked well for things like the initial versions of CUBLAS or CUFFT. But after a while, NVIDIA started to mix the pointer types of function parameters. And there is no way to *technically* distinguish between the pointer types. 

This made it necessary to have something like this in my code generator:

    Predicate<Parameter> isReadOnlyHostMemory = 
        ParameterPredicates.parameterMatches("cudnnSetTensor", "value");

Yes, this is something that I'm **really NOT** proud of :-/ because it's brittle: It worked up to CUDNN 7.0, where the `cudnnSetTensor` function was declared like that:

    /* Set all data points of a tensor to a given value : srcDest = value */
    cudnnStatus_t CUDNNWINAPI cudnnSetTensor( cudnnHandle_t                   handle,
                                              const cudnnTensorDescriptor_t   srcDestDesc,
                                              void                           *srcDestData,
                                              const void                     *value
                                             );

But in CUDNN 8.0+, it was declared like that:

    /* Set all values of a tensor to a given value : y[i] = value[0] */
    cudnnStatus_t CUDNNWINAPI cudnnSetTensor(
                                    cudnnHandle_t                       handle,
                                    const cudnnTensorDescriptor_t       yDesc,
                                    void                               *y,
                                    const void                         *valuePtr );

Usually, I'm aware of that, and try to keep an eye on such subtle renamings, but I missed that one. 

Technically, I *could* treat pointers to host- and device memory equally at this point, but handling pointers to host memory is far more complicated (in terms of "amount of code" and efficiency). So I usually try to only generate the host-pointer-handling code (which involves the `PointerData`) when it is necessary. But in doubt, I'll probably go for the more generic approach, because otherwise, the only way to distinguish host- and device pointers is via the documentation, and when there are many functions with mixed parameter types, this may become infeasible.

**Comment by blueberry** [comment]
Thank you, @jcuda 

Is there a rough estimate on when you'll find some time to work on the 10.2 release? I implemented a temporary workaround through my own kernels so I am not in a hurry, but it would be nice to be able to plan for the 10.2 transition.

**Comment by jcuda** [comment]
There are some possible tasks for this and next week that I'm not yet sure about, but I already had a look at the release notes, and it seems that there are not sooo many API changes, so it should be possible to squeeze in an update by end of next week. (Maybe mid of next week, if I manage to allocate some time during the weekend).

BTW: Do you have an idea how to address the issue about the libstdc++ version that was mentioned in https://forum.byte-welt.net/t/why-jcuda-10-1-requires-usr-local-bfm-lib64-libstdc-so-6-version-cxxabi-1-3-8/21225/5 ? I'd like to avoid having to create a "10.2.1" version shortly after the "10.2.0" release...

**Comment by blueberry** [comment]
I really don't. I use Arch Linux, which is a rolling release that updates dependencies often, and is not explicitly supported by Nvidia and other major vendors. The generic CUDA/cuDNN installer works, but often needs some manual hacking of dependencies, one of them being asking for a slightly older GCC.

Supporting an ancient GCC is, I suppose, possible, but I often needed to poke around even to make Nvidia's binaries work with GCC 6, so I don't know how to support GCC 4.X.X. I can find my way through various C++ shenanigans, but I can't say I'm an expert, especially when CMake gets involved.

In short, I would gladly build a Linux binary with an older libstdc++ that follows the minimal Red Hat requirements, but I don't know how to do that. C++ is not my primary (nor in top 5 :) language. Big companies can afford a lot of overhead to support very old setups.

**Comment by blueberry** [comment]
@jcuda Maybe the best solution would be that I provide the Linux binaries for the RC right away, and the people who need the older version compatible with Red Hat defaults build it with their version for the final version when they can (or right away :)

**Comment by blueberry** [comment]
This works now! Thank you, @jcuda!
